### PR TITLE
Fix `hiopVectorRajaPar::copyToStartingAt_w_pattern`

### PR DESCRIPTION
**Pull request by nychiang**
Fix `hiopVectorRajaPar::copyToStartingAt_w_pattern`, for running hiop sparse on device.
Add corresponding unit test.

The current implementation looks ugly, but it works.
Do we have a better way to implement the following function?
image

### CODE REVIEW

**Comment by cnpetra** [comment]
I cannot think of anything else than the inclusive scan. An optimization would be to store the cumsum indexes and reuse them when the method is called again (it is called twice per vector per iteration).

**Comment by nychiang** [comment]
@cnpetra another review please. 👍

File: src/LinAlg/hiopVectorRajaPar.cpp [hunk 1]
```diff
@@ -563,34 +564,58 @@ void hiopVectorRajaPar::copyToStarting(hiopVector& vec, int start_index_in_dest)
...
+    RAJA_LAMBDA(RAJA::Index_type i)
     {
-      assert(id[i] == zero || id[i] == one);
-      if(id[i] == one){
-        vd[start_index_in_dest+find_nnz] = dd[i];
-        find_nnz++;
+      if(nnz_in_row[i] != nnz_in_row[i-1]){
+        assert(nnz_in_row[i] == nnz_in_row[i-1] + 1);
+        index_type idx_dest = nnz_in_row[i-1] + start_index_in_dest;
+        vd[idx_dest] = dd[i-1];
```
Comment (cnpetra) [inline src/LinAlg/hiopVectorRajaPar.cpp:615]: I would add an `assert` to check `idx_dest` is less than `vec.n_local_`

File: src/LinAlg/hiopVectorRajaPar.hpp [hunk 2]
```diff
@@ -153,7 +153,10 @@ class hiopVectorRajaPar : public hiopVector
   virtual void copyToStarting(int start_index_in_src, hiopVector& v) const;
   /* Copy 'this' to v starting at start_index in 'v'. */
   virtual void copyToStarting(hiopVector& v, int start_index_in_dest) const;
-  virtual void copyToStartingAt_w_pattern(hiopVector& v, int start_index_in_dest, const hiopVector& ix) const;
+  virtual void copyToStartingAt_w_pattern(hiopVector& v,
+                                          int start_index_in_dest,
+                                          const hiopVector& ix,
+                                          hiopVectorInt* idx_map = nullptr) const;
```
Comment (cnpetra) [inline src/LinAlg/hiopVectorRajaPar.hpp:159]: I think having `idx_map` as a member in the vector class would offer better encapsulation. A good name would be `idx_cumsum_`

**Review by cnpetra — CHANGES_REQUESTED** [review]

File: src/LinAlg/hiopVectorRajaPar.hpp (hunk 2 above)
Comment (nychiang) [reply src/LinAlg/hiopVectorRajaPar.hpp]: the reason I didn't implemented it like the way you mentioned, is that `idx_cumsum_` may need to be updated in many functions. For example, in `copyTo` and `copyFrom`. See the latest commit for example. (note that I haven't reset `idx_map` in all the functions yet)

**Comment by nychiang** [comment]
Use mutable variable.
@cnpetra

**Review by cnpetra — APPROVED** [review]

**Pull request merged**